src/topology/topological_fiber_bundle.lean

```diff
@@ -0,0 +1,398 @@
+are in fact no dependent type difficulties here!
+
+For this construction, we should thus choose for each `x` one specific trivialization around it. We
+include this choice in the definition of the fiber bundle, as it makes some constructions more
+functorial and it is a nice way to say that the trivializations cover the whole space B.
+
+With this definition, the type of the fiber bundle is just `Σ x : B, F`, but the topology is not the
+product one.
+
+We also take the indexing type (indexing all the trivializations) as a parameter to the bundle:
+it could always be taken as a subtype of all the maps from open subsets of B to continuous maps
+of F, but in practice it will sometimes be something else. For instance, on a manifold, one will use
+the set of charts as a good parameterization for the trivializations of the tangent bundle. Or for
+the pullback of a fiber bundle the indexing type will be the same as for the initial bundle.
+
+## Tags
+Fiber bundle, topological bundle, vector bundle, local trivialization, structure group
+-/
+
+variables {ι : Type*} {B : Type*} {F : Type*}
+
+open topological_space set
+
+section is_topological_fiber_bundle
+
+variables {Z : Type*} [topological_space B] [topological_space Z]
+  [topological_space F] (proj : Z → B)
+
+variable (F)
+
+/-- A topological fiber bundle with fiber F over a base B is a space projecting on B for which the
+fibers are all homeomorphic to F, such that the local situation around each point is a direct
+product. We do not register the continuity of `proj` in the definition, as it follows from the
+other conditions. -/
+def is_topological_fiber_bundle : Prop :=
+  ∀x:B, ∃e: local_homeomorph Z (B × F), ∃o:set B, x ∈ o ∧ is_open o ∧ e.source = proj⁻¹' o ∧
+  e.target = set.prod o univ ∧ (∀p ∈ e.source, (e.to_fun p).1 = proj p)
+
+variables {F} {proj}
+
+/-- The projection from a topological fiber bundle to its base is continuous. -/
+lemma is_topological_fiber_bundle.continuous_proj (h : is_topological_fiber_bundle F proj) :
+  continuous proj :=
+begin
+  suffices : ∀x, continuous_at proj x, by rwa continuous_iff_continuous_at,
+  assume x s hs,
+  rw mem_nhds_sets_iff at hs,
+  rcases hs with ⟨t, st, t_open, xt⟩,
+  rcases h (proj x) with ⟨e, o, xo, o_open, e_source, e_target, hproj⟩,
+  let u := o ∩ t,
+  have u_open : is_open u := is_open_inter o_open t_open,
+  have xu : proj x ∈ u := ⟨xo, xt⟩,
+  /- u is a small enough open neighborhood of `proj x`, contained in a trivialization domain o.
+    One should show that its preimage is open. -/
+  suffices : is_open (proj ⁻¹' u),
+  { have : proj ⁻¹' u ∈ nhds x := mem_nhds_sets this xu,
+    apply filter.mem_sets_of_superset this,
+    exact preimage_mono (subset.trans (inter_subset_right _ _) st) },
+  -- to do this, rewrite `proj ⁻¹' u` in terms of the trivialization, and use its continuity.
+  have : proj ⁻¹' u = e.source ∩ e.to_fun ⁻¹' (set.prod u univ),
```

Comment:
Minor nitpick: Can you change this into a `rw show` instead of `have : .... rw this`? That seems to do exactly what you explain in the comment.

Reply:
I don't like `rw show` because I think it is hard to read unless you are really familiar with this idiom: it is doing two things at the same time. I long for the simplicity of Isabelle proofs which are just sequences of `have : ...`, so I try to keep the complexity as low as possible to help a potential reader. (Amazingly, I don't even know how to do a rewrite in Isabelle, I never needed it!)

Reply:
Sure. Fine with me.